src/hotspot/share/runtime/os.cpp

```diff
@@ -705,21 +705,61 @@ void* os::realloc(void *memblock, size_t size, MEMFLAGS memflags, const NativeCa
     return NULL;
   }
 
-  const size_t new_outer_size = size + MemTracker::overhead_per_malloc();
+  if (MemTracker::enabled()) {
+    // NMT realloc handling
 
-  // If NMT is enabled, this checks for heap overwrites, then de-accounts the old block.
-  void* const old_outer_ptr = MemTracker::record_free(memblock);
+    const size_t new_outer_size = size + MemTracker::overhead_per_malloc();
 
-  ALLOW_C_FUNCTION(::realloc, void* const new_outer_ptr = ::realloc(old_outer_ptr, new_outer_size);)
-  if (new_outer_ptr == NULL) {
-    return NULL;
-  }
+    // Handle size overflow.
+    if (new_outer_size < size) {
+      return NULL;
+    }
+
+    const size_t old_size = MallocTracker::malloc_header(memblock)->size();
+
+    // De-account the old block from NMT *before* calling the real realloc(3) since it
+    // may invalidate old block including its header. This will also perform integrity checks
+    // on the old block (e.g. overwriters) and mark the old header as dead.
+    void* const old_outer_ptr = MemTracker::record_free(memblock);
+
+    // the real realloc
+    ALLOW_C_FUNCTION(::realloc, void* const new_outer_ptr = ::realloc(old_outer_ptr, new_outer_size);)
+
+    if (new_outer_ptr == NULL) {
+      // If realloc(3) failed, the old block still exists. We must re-instantiate the old
+      // NMT header then, since we marked it dead already. Otherwise subsequent os::realloc()
+      // or os::free() calls would trigger block integrity asserts.
+      void* p = MemTracker::record_malloc(old_outer_ptr, old_size, memflags, stack);
```

Comment:
I'm not familiar with NMT details - are `memflags` and `stack` guaranteed to be the same as the original allocation?

Reply:
Not guaranteed, but it can happen.

`memflags` can theoretically be different, the memory would then just be re-booked to the new category. In practice, nobody does this.

`stack` can differ. Memory is accounted per { stack, flags } tupel in a hash table. Only matters for nmt "details" level though.

Reply:
Don't you need to restore things exactly the way they were?

Reply:
Shoot, you are right. memflags would be easy, but stack not so much. 

This may be difficult to fix. Note that the only negative effect this has is that in the very low probability case that realloc fails  *and* we run with NMT=detail, the original allocation is re-booked under the wrong stack.

I'll check how difficult this is to fix.